spec/p2p/reactor-survey.md

```diff
@@ -0,0 +1,247 @@
+at most one block". Note that these are just simple examples that are neither ensured
+by current implementation nor are actually required by the system (e.g., typically we 
+do not care about "all" transaction but rather "all valid"). 
+Which of such requirements are actually ensured by the 
+system depends on the guarantees by the concerned parts of the system.
+
+The mempool is a distributed pool of "pending" transactions.
+A pending transaction is a "valid" transaction that has been submitted by a
+client of the blockchain but has not yet been committed to the blockchain.
+The mempool is thus fed with client transactions,
+that a priori can be submitted to any node in the network.
+And it is consumed by the consensus protocol, more specifically by validator nodes,
+which retrieve from the mempool transactions to be included in proposed blocks.
+
+> Note that in this document, we want to capture (an overapproximation of) the requirements the mempool puts on p2p. The notions of "pending" and "valid" seem mostly to be conditions that the mempool can use to drop transactions (and exempt them from beeing spread over the network). Thus for the purpose of this document we will largely ignore them (they should eventually be addressed [here](https://github.com/cometbft/cometbft/issues/223)). 
+
+As 
+- full nodes add transactions to the mempool following client requests, and
+- validators read transactions from the mempool to compose proposed blocks,
+
+the mentioned end-to-end requirements on transactions translate into requirements 
+on the mempool, e.g., every transaction that is submitted to a full node should 
+eventually be present at the mempool of a validator/proposer. 
+
+For the purpose of this document (specifying what the mempool needs from p2p), 
+we do not strive to precisely state the end-to-end requirements. Rather, we 
+will show that the arguably strongest requirement on the mempool do translate roughly
+into requirements on the p2p layer that are similar to those of the
+consensus reactor.
+
+- end-to-end requirement: "every transaction submitted should be eventually put into a block"
+> leads to
+- requirement for consensus: for every submitted transaction from some height on, 
+the transaction is proposed by every correct proposer in every height until it is committed
+into a block (assuming infinitely often correct proposers get their proposal through)  
+> leads to
+- current solution to address this requirement: 
+    - the mempool of correct full nodes is organized as a list. 
+    This list can be understood as a shared data structure with relaxed semantics, that is,
+    not all nodes need to agree on the order of the elements, but we have a property that 
+    approximately says "if at time *t* a transaction tx1 is in the list of **all** nodes, and a 
+    client submits transaction tx2 after time t, then it is never the case the tx2 is before
+    tx1 in a list of a correct validator.
```

Comment:
This property is not provided by the current spec or implementation. It is not possible to guarantee in a weakly-connected, eventually-consistent, gossip-based network as is defined by the p2p layer.

Reply:
I think this may boil down to the definition of a "correct" node.  If "correct" means that a node will reliably receive any message gossiped from some other "correct" node within some time bound, then "correct" is relative to an observer, it has no well-defined value in general, and it can't be used to assert stuff like what is asserted here.

If you want to define "correct" in this way, and use that definition to say things like "it is never the case that tx2 is before tx1 among correct nodes", then you're defining a node that suffers any kind of network error to be incorrect, which is unknowable by the protocol as evaluated at any given node.

It is entirely possible that a transaction arrives in the mempool of only (say) 70% of the validators in a network, and if the proposer is one of those 70% of validators that transaction gets put into a proposed block, and that block is validated and committed, all without that transaction ever entering the mempool of (say) 30% of the validators of that network. Those 30% validators can fail to receive the gossip messages that would put the transaction in their mempool, while at the same time successfully receiving (and approving) the consensus messages required to commit a block which included that transaction to the chain.

So "if at time t1 a transaction tx1 is in the list of all nodes" is never possible to assume, because there is no situation where any given transaction is guaranteed to be known by all validators, unless it is put through consensus (and even then it is only guaranteed to be known to the approving quorum); "and a client submits transaction tx2 after time t1" is not something that can even be evaluated, as times t1 and t2 are not reliably comparable, and even if they were that comparison can only be made by nodes which know about both tx1 and tx2, which are not guaranteed to be a quorum (or even a majority) of nodes at any time; and finally if "it is never the case that tx2 is before tx1 in a list of a correct validator" defines order as time order, which is not necessarily true (app side mempools, and alternative prepare proposal hooks, can re-order mempool tx order by any criteria) even if time is reliable (which it isn't, afaiu).